server/src/main/java/org/apache/iotdb/db/metadata/idtable/deviceID/DeviceIDFactory.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.metadata.idtable.deviceID;
+
+import org.apache.iotdb.commons.path.PartialPath;
+import org.apache.iotdb.commons.utils.TestOnly;
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+
+import java.util.function.Function;
+
+/** factory to build device id according to configured algorithm */
+public class DeviceIDFactory {
+
+  /** used to obtain a IDeviceID instance in the query operation */
+  Function<String, IDeviceID> getDeviceIDFunction;
+
+  /** used to obtain a IDeviceID instance in the insert operation */
+  Function<String, IDeviceID> getDeviceIDWithAutoCreateFunction;
+
+  /** used to obtain a IDeviceID instance in the system restart */
+  Function<String[], IDeviceID> getDeviceIDWithRecoverFunction;
```

Comment:
Consider encapsulate this three function into one strategy class. Each strategy takes care of there own init, reset and clear etc. behaviors.